examples/LaravelSampleApp/app/Http/Controllers/GoogleAdsApiController.php

```diff
@@ -76,52 +78,112 @@ public function showReportAction(
-            $response = $googleAdsClient->getGoogleAdsServiceClient()->search(
-                $customerId,
-                $query
+                $reportRange,
+                self::RESULTS_LIMIT
             );
 
-            // Fetches all the results.
-            $results = [];
-            foreach ($response->iterateAllElements() as $googleAdsRow) {
-                /** @var GoogleAdsRow $googleAdsRow */
-                // Converts each result as a Plain Old PHP Object (POPO) using JSON.
-                $results[] = json_decode($googleAdsRow->serializeToJsonString(), true);
-            }
-            $collection = collect($results);
+            // Initializes the list of page tokens. Page tokens are used to request specific pages
+            // of results from the API. They are especially useful to optimize navigation between
+            // pages as there is no need to cache all the results before displaying.
+            // More details can be found here:
+            // https://developers.google.com/google-ads/api/docs/reporting/paging.
+            //
+            // The first page's token is always an empty string.
+            $pageTokens = [''];
 
             // Updates the session with all the information that is necessary to process any
-            // following GET requests (report result pages).
+            // future requests (report result pages).
+            $request->session()->put('customerId', $customerId);
             $request->session()->put('selectedFields', $selectedFields);
             $request->session()->put('entriesPerPage', $entriesPerPage);
-            $request->session()->put('collection', $collection);
+            $request->session()->put('query', $query);
+            $request->session()->put('pageTokens', $pageTokens);
         } else {
-            // Loads from the session all the information that is necessary to process the GET
-            // request (report result page).
+            // Loads from the session all the information that is necessary to process any
+            // requests (report result page).
+            $customerId = $request->session()->get('customerId');
             $selectedFields = $request->session()->get('selectedFields');
             $entriesPerPage = $request->session()->get('entriesPerPage');
-            $collection = $request->session()->get('collection');
+            $query = $request->session()->get('query');
+            $pageTokens = $request->session()->get('pageTokens');
         }
 
+        // Determines the number of the page to load (the first one by default).
         $pageNo = $request->input('page') ?: 1;
 
-        // Creates a length aware paginator to supply a given page of results for the view, based on
-        // the collection of results, the page number and the specified number of entries per page.
+        // Fetches pages in sequence to find the page token of the requested page if unknown.
+        while (count($pageTokens) < $pageNo) {
```

Comment:
I have a trouble understanding the purpose of this loop, partly because of the if clause at the end and the `end($pageTokens)` as well.

Can you elaborate more on what to expect at the beginning of this loop too?
The loop seems to go backward? From the end of the `$pageTokens` array? 
Is there any particular reason why we do like this instead of going forward?

Reply:
Or is this a kind of pre-processing?